Technical/Logging/The-Tazama-Logging-Framework.md

```diff
@@ -1 +1,405 @@
+```
+#### Usage
+
+You may specify a `LOGSTASH_LEVEL` environment variable in your processor to control your logs. The logging system supports multiple levels of severity, listed in ascending order of importance: `trace`, `debug`, `info`, `warn`, `error`, and `fatal`. When configuring the logging level, every log event with that level or a higher severity will be logged. For example, if you set the logging level to `info`, all log events at `info`, `warn`, `error`, and `fatal` levels will be logged. This ensures that important information and higher severity events are captured while filtering out less critical details.
+
+After initialising your logger, you may use the methods available to the `LoggerService` instance.
+
+##### Logger Service 
+
+
+###### Function Signature:
+
+The custom logger function has the following signature for `trace`, `debug`, `log` and `warn` log levels:
+
+```typescript
+trace(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+debug(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+log(message: string, serviceOperation?: string, id?: string, callback?: LogCallback): void
+```
+
+- **`message` (required)**: A string that represents the log message to be recorded.
+
+- **`serviceOperation` (optional)**: Specifies the name or identifier of the service operation related to the log message. This parameter provides context about the operation being logged.
+
+- **`id` (optional)**: An optional identifier associated with the log message. Use this parameter to correlate logs related to a specific transaction or request.
+
+- **`callback` (optional)**: A callback function that can be provided to handle asynchronous logging operations or to receive notifications after logging. This parameter is optional and may not be used in all logging scenarios.
+
+Additional Levels:
+`error` and `fatal` levels are also supported but they have a custom signature:
+
+```ts
+error(message: string | Error, innerError?: unknown, serviceOperation?: string, id?: string, callback?: LogCallback): void;
+```
+
+Which allow an instance of an [`Error`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error) to be sent as an alternative of the usual `string`, for cases where you may want more information contained in the error itself instead of just the error message (`string`)
+
+
+
+##### Usage:
+
+- **`trace`**: Used for very detailed or fine-grained informational events. Example: tracing function calls.
+
+- **`debug`**: Used for debugging purposes, providing detailed information for diagnosing issues.
+
+- **`warn`**: Indicates potential issues that should be monitored or investigated.
+- **`error`**: Indicates a significant problem that might affect the functionality of the application but does not necessarily cause it to crash immediately.
+- **`fatal`**: a critical problem that has caused the application to abort or terminate.
+
```

Comment:
I'd recommend an additional section here with our standard and accepted best practices. When to log, when not to log, warnings about excessive logging, etc.